lib/auth/iam_join.go

```diff
@@ -0,0 +1,245 @@
+X-Teleport-Challenge: 0ezlc3usTAkXeZTcfOazUq0BGrRaKmb4EwODk8U7J5A
+
+Action=GetCallerIdentity&Version=2011-06-15
+*/
+func validateSTSIdentityRequest(req *http.Request, challenge string) error {
+	if req.Host != stsHost {
+		return trace.AccessDenied("sts identity request is for unknown host %q", req.Host)
+	}
+
+	if req.Method != http.MethodPost {
+		return trace.AccessDenied("sts identity request method %q does not match expected method %q", req.RequestURI, http.MethodPost)
+	}
+
+	if req.Header.Get(challengeHeaderKey) != challenge {
+		return trace.AccessDenied("sts identity request does not include challenge header or it does not match")
+	}
+
+	authHeader := req.Header.Get(aws.AuthorizationHeader)
+
+	sigV4, err := aws.ParseSigV4(authHeader)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !utils.SliceContainsStr(sigV4.SignedHeaders, normalizedChallengeHeaderKey) {
+		return trace.AccessDenied("sts identity request auth header %q does not include "+
+			normalizedChallengeHeaderKey+" as a signed header", authHeader)
+	}
+
+	body, err := aws.GetAndReplaceReqBody(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if !bytes.Equal([]byte(expectedSTSIdentityRequestBody), body) {
+		return trace.BadParameter("sts request body %q does not equal expected %q", string(body), expectedSTSIdentityRequestBody)
+	}
+
+	return nil
+}
+
+func parseSTSRequest(req []byte) (*http.Request, error) {
+	httpReq, err := http.ReadRequest(bufio.NewReader(bytes.NewReader(req)))
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Unset RequestURI and set req.URL instead (necessary quirk of sending a
+	// request parsed by http.ReadRequest). Also, force https here.
+	if httpReq.RequestURI != "/" {
+		return nil, trace.AccessDenied("unexpected sts identity request URI: %q", httpReq.RequestURI)
+	}
+	httpReq.RequestURI = ""
+	httpReq.URL = &url.URL{
+		Scheme: "https",
+		Host:   stsHost,
+	}
+	return httpReq, nil
+}
+
+type stsIdentityResponse struct {
+	GetCallerIdentityResponse struct {
+		GetCallerIdentityResult awsIdentity
+	}
+}
+
+type awsIdentity struct {
+	Account string
+	Arn     string
+}
+
+type stsClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+func executeSTSIdentityRequest(ctx context.Context, client stsClient, req *http.Request) (identity awsIdentity, err error) {
+	req = req.WithContext(ctx)
+	resp, err := client.Do(req)
+	if err != nil {
+		return identity, trace.Wrap(err)
+	}
+	defer resp.Body.Close()
+
+	body, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return identity, trace.Wrap(err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return identity, trace.AccessDenied("aws sts api returned status: %q body: %q",
+			resp.Status, body)
+	}
+
+	var identityResponse stsIdentityResponse
+	if err := json.Unmarshal(body, &identityResponse); err != nil {
+		return identity, trace.Wrap(err)
+	}
+	return identityResponse.GetCallerIdentityResponse.GetCallerIdentityResult, nil
```

Comment:
Is it worth checking if the fields on `identityResponse.GetCallerIdentityResponse.GetCallerIdentityResult` are set?

Reply:
might as well, done